README
Business-IBAN-Validator

This module provides a validator for IBANs (International Bank Account Numbers)
(ISO 13616).

The validation consists of a number of tests:

- Know country code (ISO 3166 alpha2)
- Correct length (different for each country)
- Correct pattern (different for each country)
- Correct 97-checksum

On top of that, the validator also knows which countries participate in SEPA
(Single Euro Payment Area).

The "Database" with this information is extracted from a document published by
SWIFT (Society for Worldwide Interbank Financial Telecommunication) in April
2013 (IBAN Registry Release 45).

INSTALLATION

To install this module, run the following commands:

        perl Makefile.PL
        make
        make test
        make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Business::IBAN::Validator

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-IBAN-Validator

    Search CPAN
        http://search.cpan.org/dist/Business-IBAN-Validator


COPYRIGHT AND LICENCE

Copyright (C) 2013 Abe Timmerman

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
0.07_01 2023-10-10T10:30:17+02:00 (5aca2ae => Abe Timmerman)
 - (Abe Timmerman, Tue, 10 Oct 2023 10:30:17 +0200) Autocommit for
   distribution Business-IBAN-Validator 0.07_01 (test)

0.07_00 2023-10-09T14:46:12+02:00 (aa6e39b => Abe Timmerman)
 - (Abe Timmerman, Mon, 9 Oct 2023 14:46:12 +0200) Autocommit for
   distribution Business-IBAN-Validator 0.07_00 (test)
 - (Abe Timmerman, Mon, 9 Oct 2023 14:49:41 +0200) Fix autogen-error in
   debian/changelog
 - (Abe Timmerman, Tue, 10 Oct 2023 10:24:40 +0200) FIX: 'bugtracker' is
   part of 'resources' in the META-spec
 -     The bugtracker wasn't recognised by MetaCPAN because it was in the
   wrong
 -     place (in META_ADD).
 - (Abe Timmerman, Tue, 10 Oct 2023 10:27:00 +0200) Remove the fixed
   testcount from t/00-load.t
 -     Now that the 'no warnings' test is only run for a true
 -     $ENV{AUTHOR_TESTING}, a fixed test count doesn't make sense.

0.07 2023-04-12T17:48:19+02:00 (083ace3 => Abe Timmerman)
 - (Abe Timmerman, Wed, 12 Apr 2023 17:48:19 +0200) Autocommit for
   distribution Business::IBAN::Validator 0.07 (minor)
 - (Abe Timmerman, Mon, 9 Oct 2023 13:54:31 +0200) Update for revision 95
   (April 2023)
 - (Abe Timmerman, Mon, 9 Oct 2023 14:00:53 +0200) Add missing information
   to the debian/copyright file.
 - (Abe Timmerman, Mon, 9 Oct 2023 14:20:30 +0200) Update t::Test::abeltje
   to 1.09
 - (Abe Timmerman, Mon, 9 Oct 2023 14:21:33 +0200) Regenerate the example
   tests from v95

0.06_02 2023-03-03T13:06:06+01:00 (efb0d3d => Abe Timmerman)
 - (Abe Timmerman, Fri, 3 Mar 2023 13:06:06 +0100) Autocommit for
   distribution Business::IBAN::Validator 0.06_02 (test)

0.06_01 2023-03-03T12:59:19+01:00 (beea72c => Abe Timmerman)
 - (Abe Timmerman, Fri, 3 Mar 2023 12:59:19 +0100) Autocommit for
   distribution Business::IBAN::Validator 0.06_01 (test)

0.06 2015-02-04T16:11:44+01:00 (b26745d => Abe Timmerman)
 - (Abe Timmerman, Wed, 4 Feb 2015 16:11:44 +0100) Autocommit for
   distribution Business::IBAN::Validator 0.06
 - (Abe Timmerman, Wed, 1 Mar 2023 10:07:48 +0100) Introduce
   t::Test::abeltje 1.08
 -     Do some cosmetics on the files in the distro.
 - (Abe Timmerman, Wed, 1 Mar 2023 20:13:30 +0100) Update to SWIFT document
   version 93: February 2023
 -     The SWIFT website now also supplies a new document in txt format
 -     (tab-separated, with a heading per line) that we now use to generate
   the
 -     Database-module and the auto-example-test.
 -     Split the util-functions into their own module for multiple use and
 -     better maintainance.
 - (Abe Timmerman, Wed, 1 Mar 2023 20:18:10 +0100) Add the new SWIFT
   documents to the repository
 -     We used these for generation of the Business::IBAN::Database module
   and
 -     the t/020-iban-auto.t test.
 - (Abe Timmerman, Fri, 3 Mar 2023 12:57:25 +0100) Make VERSION a string so
   we can have beta-versions

0.05 2013-11-12T11:53:51+01:00 (b083599 => abeltje)
 - (abeltje, Tue, 12 Nov 2013 11:53:51 +0100) Autocommit for distribution
   Business::IBAN::Validator 0.05
 - (Tekki, Wed, 4 Feb 2015 15:34:05 +0100) Update to registry version 54
   (january 2015).
 - (Abe Timmerman, Wed, 4 Feb 2015 16:00:08 +0100) Merge pull request #1
   from Tekki/master
 -     Update to registry version 54 (january 2015) by Tekki.

0.04 2013-11-04T23:18:34+01:00 (aa5f448 => abeltje)
 - (abeltje, Mon, 4 Nov 2013 23:18:34 +0100) Autocommit for distribution
   Business::IBAN::Validator 0.04
 - (abeltje, Tue, 12 Nov 2013 11:51:11 +0100) RT-90271 - Peter Heimann:
   Update to IBAN Registry Release 46 November 2013
 -     Country name capitalization should be unified
 -     Country Qatar added in SWIFT IBAN Registry, Release 46, November
   2013
 -     Added the IBAN Registry document Release 46
 -     Remove support/ dir from the CPAN-distribution


Makefile.PL
#! perl -w
use strict;
use 5.008;

use ExtUtils::MakeMaker 6.55_01;

my $testfiles = 't/*.t';
$testfiles = "xt/*.t $testfiles" if $ENV{AUTHOR_TESTING};

WriteMakefile(
    NAME          => 'Business::IBAN::Validator',
    VERSION_FROM  => 'lib/Business/IBAN/Validator.pm',
    ABSTRACT_FROM => 'lib/Business/IBAN/Validator.pm',
    AUTHOR        => 'Abe Timmerman <abeltje@cpan.org>',

    PREREQ_PM => {
        'Hash::Util' => 0, # It's core, but still...
    },

    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 6.56,
    },

    BUILD_REQUIRES => {
        'Test::Simple'   => 0.88,    # done_testing()
        'Test::Fatal'    => 0,
        'Test::Warnings' => 0,
    },

    MIN_PERL_VERSION => 5.008,
    META_ADD => {
        'meta-spec' => {
            version => 2,
        },
        license => [ 'perl_5' ],
        'resources' => {
            repository => {
                type => 'git',
                url  => 'https://github.com/abeltje/Business-IBAN-Validator.git',
                web  => 'https://github.com/abeltje/Business-IBAN-Validator',
            },
        },
        bugtracker => {
            web => 'https://github.com/abeltje/Business-IBAN-Validator/issues',
        },
    },

    test => { TESTS => $testfiles },
);


lib/Business/IBAN/Validator.pm
package Business::IBAN::Validator;
use warnings;
use strict;

our $VERSION = "0.07_01";

use Hash::Util qw/unlock_hash lock_hash/;
use Business::IBAN::Database;

sub new {
    my $class = shift;

    my $db = iban_db();
    unlock_hash(%$db);
    my $self =  bless $db, $class;
    lock_hash(%$self);
    return $self;
}

sub validate {
    my $self = shift;
    my ($iban) = @_;

    (my $to_check = $iban) =~ s/\s+//g;
    my $iso3166a2 = uc(substr($to_check, 0, 2));

    if (not exists($self->{$iso3166a2})) {
        die "'$iso3166a2' is not an IBAN country code.\n";
    }

    my $iban_info = $self->{$iso3166a2};
    if (length($to_check) != $iban_info->{iban_length}) {
        die(
            sprintf(
                "'%s' has incorrect length %d (expected %d for %s).\n",
                $iban,
                length($to_check),
                $iban_info->{iban_length},
                $iban_info->{country}
            )
        );
    }

    if ($to_check !~ $iban_info->{iban_structure}) {
        die(
            sprintf(
                "'%s' does not match the pattern '%s'for %s.\n",
                $iban,
                $iban_info->{pattern},
                $iban_info->{country}
            )
        );
    }

    if ( mod97(numify_iban($to_check)) != 1) {
        die "'$iban' does not comply with the 97-check.\n";
    }

    return 1;
}

sub is_sepa {
    my $self = shift;
    my ($iban) = @_;
    (my $to_check = $iban) =~ s/\s+//g;
    my $iso3166a2 = uc(substr($to_check, 0, 2));

    if (not exists($self->{$iso3166a2})) {
        die "'$iso3166a2' is not an IBAN country code.\n";
    }

    return $self->{$iso3166a2}{is_sepa};
}

sub DESTROY { }

sub AUTOLOAD {
    my $self = shift;
    (my $cc =  our $AUTOLOAD) =~ s/.*:://;
    if (!exists $self->{uc $cc}) {
        require Carp;
        Carp::croak("'$cc' is not a valid IBAN country code.");
    }
    return $self->{uc $cc};
}

1;

=head1 NAME

Business::IBAN::Validator - A validator for the structure of IBANs.

=head1 SYNOPSIS

    use Business::IBAN::Validator;
    my $v = Business::IBAN::Validator->new;
    while (1) {
        print 'Enter IBAN: ';
        chomp(my $input = <>);
        last if !$input;
        eval { $v->validate($input) };
        if ($@) {
            print "Not ok: $@";
        }
        else {
            print "OK\n";
        }
    }

=head1 DESCRIPTION

This module does a series of checks on an IBAN:

=over

=item Country code

=item Length

=item Pattern

=item 97-Check

=back

=head2 $v = Business::IBAN::Validator->new()

Return an IBAN validator object.

=head2 $v->validate($iban)

Perform a series of checks, and die() as soon as one fails.

Return 1 on success.

=head2 $v->is_sepa($iban)

Return the SEPA status of the country (as denoted by the first two letters).

=head1 COPYRIGHT

E<copy> MMXIII-MMXV - Abe Timmerman <abeltje@cpan.org>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut


lib/Business/IBAN/Database.pm
package Business::IBAN::Database;
use warnings;
use strict;

use Hash::Util 'lock_hash';
use Exporter 'import';
our @EXPORT = qw/iban_db numify_iban mod97/;

my %iban_db = (
  'AD' => {
    'country' => 'Andorra',
    'iban_length' => '24',
    'iban_structure' => 'AD[0-9]{2}[0-9]{4}[0-9]{4}[A-Za-z0-9]{12}',
    'is_sepa' => 0,
    'pattern' => 'AD2!n4!n4!n12!c'
  },
  'AE' => {
    'country' => 'United Arab Emirates',
    'iban_length' => '23',
    'iban_structure' => 'AE[0-9]{2}[0-9]{3}[0-9]{16}',
    'is_sepa' => 0,
    'pattern' => 'AE2!n3!n16!n'
  },
  'AL' => {
    'country' => 'Albania',
    'iban_length' => '28',
    'iban_structure' => 'AL[0-9]{2}[0-9]{8}[A-Za-z0-9]{16}',
    'is_sepa' => 0,
    'pattern' => 'AL2!n8!n16!c'
  },
  'AT' => {
    'country' => 'Austria',
    'iban_length' => '20',
    'iban_structure' => 'AT[0-9]{2}[0-9]{5}[0-9]{11}',
    'is_sepa' => 1,
    'pattern' => 'AT2!n5!n11!n'
  },
  'AZ' => {
    'country' => 'Republic of Azerbaijan',
    'iban_length' => '28',
    'iban_structure' => 'AZ[0-9]{2}[A-Z]{4}[A-Za-z0-9]{20}',
    'is_sepa' => 0,
    'pattern' => 'AZ2!n4!a20!c'
  },
  'BA' => {
    'country' => 'Bosnia and Herzegovina',
    'iban_length' => '20',
    'iban_structure' => 'BA[0-9]{2}[0-9]{3}[0-9]{3}[0-9]{8}[0-9]{2}',
    'is_sepa' => 0,
    'pattern' => 'BA2!n3!n3!n8!n2!n'
  },
  'BE' => {
    'country' => 'Belgium',
    'iban_length' => '16',
    'iban_structure' => 'BE[0-9]{2}[0-9]{3}[0-9]{7}[0-9]{2}',
    'is_sepa' => 1,
    'pattern' => 'BE2!n3!n7!n2!n'
  },
  'BG' => {
    'country' => 'Bulgaria',
    'iban_length' => '22',
    'iban_structure' => 'BG[0-9]{2}[A-Z]{4}[0-9]{4}[0-9]{2}[A-Za-z0-9]{8}',
    'is_sepa' => 1,
    'pattern' => 'BG2!n4!a4!n2!n8!c'
  },
  'BH' => {
    'country' => 'Kingdom of Bahrain',
    'iban_length' => '22',
    'iban_structure' => 'BH[0-9]{2}[A-Z]{4}[A-Za-z0-9]{14}',
    'is_sepa' => 0,
    'pattern' => 'BH2!n4!a14!c'
  },
  'BR' => {
    'country' => 'Brazil',
    'iban_length' => '29',
    'iban_structure' => 'BR[0-9]{2}[0-9]{8}[0-9]{5}[0-9]{10}[A-Z]{1}[A-Za-z0-9]{1}',
    'is_sepa' => 0,
    'pattern' => 'BR2!n8!n5!n10!n1!a1!c'
  },
  'CH' => {
    'country' => 'Switzerland',
    'iban_length' => '21',
    'iban_structure' => 'CH[0-9]{2}[0-9]{5}[A-Za-z0-9]{12}',
    'is_sepa' => 1,
    'pattern' => 'CH2!n5!n12!c'
  },
  'CR' => {
    'country' => 'Costa Rica',
    'iban_length' => '21',
    'iban_structure' => 'CR[0-9]{2}[0-9]{3}[0-9]{14}',
    'is_sepa' => 0,
    'pattern' => 'CR2!n3!n14!n'
  },
  'CY' => {
    'country' => 'Cyprus',
    'iban_length' => '28',
    'iban_structure' => 'CY[0-9]{2}[0-9]{3}[0-9]{5}[A-Za-z0-9]{16}',
    'is_sepa' => 1,
    'pattern' => 'CY2!n3!n5!n16!c'
  },
  'CZ' => {
    'country' => 'Czech Republic',
    'iban_length' => '24',
    'iban_structure' => 'CZ[0-9]{2}[0-9]{4}[0-9]{6}[0-9]{10}',
    'is_sepa' => 1,
    'pattern' => 'CZ2!n4!n6!n10!n'
  },
  'DE' => {
    'country' => 'Germany',
    'iban_length' => '22',
    'iban_structure' => 'DE[0-9]{2}[0-9]{8}[0-9]{10}',
    'is_sepa' => 1,
    'pattern' => 'DE2!n8!n10!n'
  },
  'DK' => {
    'country' => 'Denmark',
    'iban_length' => '18',
    'iban_structure' => 'DK[0-9]{2}[0-9]{4}[0-9]{9}[0-9]{1}',
    'is_sepa' => 1,
    'pattern' => 'DK2!n4!n9!n1!n FO2!n4!n9!n1!n GL2!n4!n9!n1!n'
  },
  'DO' => {
    'country' => 'Dominican Republic',
    'iban_length' => '28',
    'iban_structure' => 'DO[0-9]{2}[A-Za-z0-9]{4}[0-9]{20}',
    'is_sepa' => 0,
    'pattern' => 'DO2!n4!c20!n'
  },
  'EE' => {
    'country' => 'Estonia',
    'iban_length' => '20',
    'iban_structure' => 'EE[0-9]{2}[0-9]{2}[0-9]{2}[0-9]{11}[0-9]{1}',
    'is_sepa' => 1,
    'pattern' => 'EE2!n2!n2!n11!n1!n'
  },
  'ES' => {
    'country' => 'Spain',
    'iban_length' => '24',
    'iban_structure' => 'ES[0-9]{2}[0-9]{4}[0-9]{4}[0-9]{1}[0-9]{1}[0-9]{10}',
    'is_sepa' => 1,
    'pattern' => 'ES2!n4!n4!n1!n1!n10!n'
  },
  'FI' => {
    'country' => 'Finland',
    'iban_length' => '18',
    'iban_structure' => 'FI[0-9]{2}[0-9]{6}[0-9]{7}[0-9]{1}',
    'is_sepa' => 1,
    'pattern' => 'FI2!n6!n7!n1!n'
  },
  'FO' => {
    'country' => 'Denmark',
    'iban_length' => '18',
    'iban_structure' => 'FO[0-9]{2}[0-9]{4}[0-9]{9}[0-9]{1}',
    'is_sepa' => 1,
    'pattern' => 'DK2!n4!n9!n1!n FO2!n4!n9!n1!n GL2!n4!n9!n1!n'
  },
  'FR' => {
    'country' => 'France',
    'iban_length' => '27',
    'iban_structure' => 'FR[0-9]{2}[0-9]{5}[0-9]{5}[A-Za-z0-9]{11}[0-9]{2}',
    'is_sepa' => 1,
    'pattern' => 'FR2!n5!n5!n11!c2!n'
  },
  'GB' => {
    'country' => 'United Kingdom',
    'iban_length' => '22',
    'iban_structure' => 'GB[0-9]{2}[A-Z]{4}[0-9]{6}[0-9]{8}',
    'is_sepa' => 1,
    'pattern' => 'GB2!n4!a6!n8!n'
  },
  'GE' => {
    'country' => 'Georgia',
    'iban_length' => '22',
    'iban_structure' => 'GE[0-9]{2}[A-Z]{2}[0-9]{16}',
    'is_sepa' => 0,
    'pattern' => 'GE2!n2!a16!n'
  },
  'GI' => {
    'country' => 'Gibraltar',
    'iban_length' => '23',
    'iban_structure' => 'GI[0-9]{2}[A-Z]{4}[A-Za-z0-9]{15}',
    'is_sepa' => 1,
    'pattern' => 'GI2!n4!a15!c'
  },
  'GL' => {
    'country' => 'Denmark',
    'iban_length' => '18',
    'iban_structure' => 'GL[0-9]{2}[0-9]{4}[0-9]{9}[0-9]{1}',
    'is_sepa' => 1,
    'pattern' => 'DK2!n4!n9!n1!n FO2!n4!n9!n1!n GL2!n4!n9!n1!n'
  },
  'GR' => {
    'country' => 'Greece',
    'iban_length' => '27',
    'iban_structure' => 'GR[0-9]{2}[0-9]{3}[0-9]{4}[A-Za-z0-9]{16}',
    'is_sepa' => 1,
    'pattern' => 'GR2!n3!n4!n16!c'
  },
  'GT' => {
    'country' => 'Guatemala',
    'iban_length' => '28',
    'iban_structure' => 'GT[0-9]{2}[A-Za-z0-9]{4}[A-Za-z0-9]{20}',
    'is_sepa' => 0,
    'pattern' => 'GT2!n4!c20!c'
  },
  'HR' => {
    'country' => 'Croatia',
    'iban_length' => '21',
    'iban_structure' => 'HR[0-9]{2}[0-9]{7}[0-9]{10}',
    'is_sepa' => 0,
    'pattern' => 'HR2!n7!n10!n'
  },
  'HU' => {
    'country' => 'Hungary',
    'iban_length' => '28',
    'iban_structure' => 'HU[0-9]{2}[0-9]{3}[0-9]{4}[0-9]{1}[0-9]{15}[0-9]{1}',
    'is_sepa' => 1,
    'pattern' => 'HU2!n3!n4!n1!n15!n1!n'
  },
  'IE' => {
    'country' => 'Ireland',
    'iban_length' => '22',
    'iban_structure' => 'IE[0-9]{2}[A-Z]{4}[0-9]{6}[0-9]{8}',
    'is_sepa' => 1,
    'pattern' => 'IE2!n4!a6!n8!n'
  },
  'IL' => {
    'country' => 'Israel',
    'iban_length' => '23',
    'iban_structure' => 'IL[0-9]{2}[0-9]{3}[0-9]{3}[0-9]{13}',
    'is_sepa' => 0,
    'pattern' => 'IL2!n3!n3!n13!n'
  },
  'IS' => {
    'country' => 'Iceland',
    'iban_length' => '26',
    'iban_structure' => 'IS[0-9]{2}[0-9]{4}[0-9]{2}[0-9]{6}[0-9]{10}',
    'is_sepa' => 1,
    'pattern' => 'IS2!n4!n2!n6!n10!n'
  },
  'IT' => {
    'country' => 'Italy',
    'iban_length' => '27',
    'iban_structure' => 'IT[0-9]{2}[A-Z]{1}[0-9]{5}[0-9]{5}[A-Za-z0-9]{12}',
    'is_sepa' => 1,
    'pattern' => 'IT2!n1!a5!n5!n12!c'
  },
  'JO' => {
    'country' => 'Jordan',
    'iban_length' => '30',
    'iban_structure' => 'JO[0-9]{2}[A-Z]{4}[0-9]{4}[0-9]{18}',
    'is_sepa' => 0,
    'pattern' => 'JO2!a2!n4!n18!c' # 'JO2!a2!n4!a4!n18!c' according to registry
  },
  'KW' => {
    'country' => 'Kuwait',
    'iban_length' => '30',
    'iban_structure' => 'KW[0-9]{2}[A-Z]{4}[A-Za-z0-9]{22}',
    'is_sepa' => 0,
    'pattern' => 'KW2!n4!a22!c'
  },
  'KZ' => {
    'country' => 'Kazakhstan',
    'iban_length' => '20',
    'iban_structure' => 'KZ[0-9]{2}[0-9]{3}[A-Za-z0-9]{13}',
    'is_sepa' => 0,
    'pattern' => 'KZ2!n3!n13!c'
  },
  'LB' => {
    'country' => 'Lebanon',
    'iban_length' => '28',
    'iban_structure' => 'LB[0-9]{2}[0-9]{4}[A-Za-z0-9]{20}',
    'is_sepa' => 0,
    'pattern' => 'LB2!n4!n20!c'
  },
  'LI' => {
    'country' => 'Liechtenstein (Principality of)',
    'iban_length' => '21',
    'iban_structure' => 'LI[0-9]{2}[0-9]{5}[A-Za-z0-9]{12}',
    'is_sepa' => 1,
    'pattern' => 'LI2!n5!n12!c'
  },
  'LT' => {
    'country' => 'Lithuania',
    'iban_length' => '20',
    'iban_structure' => 'LT[0-9]{2}[0-9]{5}[0-9]{11}',
    'is_sepa' => 1,
    'pattern' => 'LT2!n5!n11!n'
  },
  'LU' => {
    'country' => 'Luxembourg',
    'iban_length' => '20',
    'iban_structure' => 'LU[0-9]{2}[0-9]{3}[A-Za-z0-9]{13}',
    'is_sepa' => 1,
    'pattern' => 'LU2!n3!n13!c'
  },
  'LV' => {
    'country' => 'Latvia',
    'iban_length' => '21',
    'iban_structure' => 'LV[0-9]{2}[A-Z]{4}[A-Za-z0-9]{13}',
    'is_sepa' => 1,
    'pattern' => 'LV2!n4!a13!c'
  },
  'MC' => {
    'country' => 'Monaco',
    'iban_length' => '27',
    'iban_structure' => 'MC[0-9]{2}[0-9]{5}[0-9]{5}[A-Za-z0-9]{11}[0-9]{2}',
    'is_sepa' => 1,
    'pattern' => 'MC2!n5!n5!n11!c2!n'
  },
  'MD' => {
    'country' => 'Republic of Moldova',
    'iban_length' => '24',
    'iban_structure' => 'MD[0-9]{2}[A-Za-z0-9]{20}',
    'is_sepa' => 0,
    'pattern' => 'MD2!n20!c'
  },
  'ME' => {
    'country' => 'Montenegro',
    'iban_length' => '22',
    'iban_structure' => 'ME[0-9]{2}[0-9]{3}[0-9]{13}[0-9]{2}',
    'is_sepa' => 0,
    'pattern' => 'ME2!n3!n13!n2!n'
  },
  'MK' => {
    'country' => 'Macedonia',
    'iban_length' => '19',
    'iban_structure' => 'MK[0-9]{2}[0-9]{3}[A-Za-z0-9]{10}[0-9]{2}',
    'is_sepa' => 0,
    'pattern' => 'MK2!n3!n10!c2!n'
  },
  'MR' => {
    'country' => 'Mauritania',
    'iban_length' => '27',
    'iban_structure' => 'MR13[0-9]{5}[0-9]{5}[0-9]{11}[0-9]{2}',
    'is_sepa' => 0,
    'pattern' => 'MR135!n5!n11!n2!n'
  },
  'MT' => {
    'country' => 'Malta',
    'iban_length' => '31',
    'iban_structure' => 'MT[0-9]{2}[A-Z]{4}[0-9]{5}[A-Za-z0-9]{18}',
    'is_sepa' => 1,
    'pattern' => 'MT2!n4!a5!n18!c'
  },
  'MU' => {
    'country' => 'Mauritius',
    'iban_length' => '30',
    'iban_structure' => 'MU[0-9]{2}[A-Z]{4}[0-9]{2}[0-9]{2}[0-9]{12}[0-9]{3}[A-Z]{3}',
    'is_sepa' => 0,
    'pattern' => 'MU2!n4!a2!n2!n12!n3!n3!a'
  },
  'NL' => {
    'country' => 'The Netherlands',
    'iban_length' => '18',
    'iban_structure' => 'NL[0-9]{2}[A-Z]{4}[0-9]{10}',
    'is_sepa' => 1,
    'pattern' => 'NL2!n4!a10!n'
  },
  'NO' => {
    'country' => 'Norway',
    'iban_length' => '15',
    'iban_structure' => 'NO[0-9]{2}[0-9]{4}[0-9]{6}[0-9]{1}',
    'is_sepa' => 1,
    'pattern' => 'NO2!n4!n6!n1!n'
  },
  'PK' => {
    'country' => 'Pakistan',
    'iban_length' => '24',
    'iban_structure' => 'PK[0-9]{2}[A-Z]{4}[A-Za-z0-9]{16}',
    'is_sepa' => 0,
    'pattern' => 'PK2!n4!a16!c'
  },
  'PL' => {
    'country' => 'Poland',
    'iban_length' => '28',
    'iban_structure' => 'PL[0-9]{2}[0-9]{8}[0-9]{16}',
    'is_sepa' => 1,
    'pattern' => 'PL2!n8!n16!n'
  },
  'PS' => {
    'country' => 'Palestine, State of',
    'iban_length' => '29',
    'iban_structure' => 'PS[0-9]{2}[A-Z]{4}[A-Za-z0-9]{21}',
    'is_sepa' => 0,
    'pattern' => 'PS2!n4!a21!c'
  },
  'PT' => {
    'country' => 'Portugal',
    'iban_length' => '25',
    'iban_structure' => 'PT[0-9]{2}[0-9]{4}[0-9]{4}[0-9]{11}[0-9]{2}',
    'is_sepa' => 1,
    'pattern' => 'PT2!n4!n4!n11!n2!n'
  },
  'QA' => {
    'country' => 'Qatar',
    'iban_length' => '29',
    'iban_structure' => 'QA[0-9]{2}[A-Z]{4}[A-Za-z0-9]{21}',
    'is_sepa' => 0,
    'pattern' => 'QA2!n4!a21!c'
  },
  'RO' => {
    'country' => 'Romania',
    'iban_length' => '24',
    'iban_structure' => 'RO[0-9]{2}[A-Z]{4}[A-Za-z0-9]{16}',
    'is_sepa' => 1,
    'pattern' => 'RO2!n4!a16!c'
  },
  'RS' => {
    'country' => 'Serbia',
    'iban_length' => '22',
    'iban_structure' => 'RS[0-9]{2}[0-9]{3}[0-9]{13}[0-9]{2}',
    'is_sepa' => 0,
    'pattern' => 'RS2!n3!n13!n2!n'
  },
  'SA' => {
    'country' => 'Saudi Arabia',
    'iban_length' => '24',
    'iban_structure' => 'SA[0-9]{2}[0-9]{2}[A-Za-z0-9]{18}',
    'is_sepa' => 0,
    'pattern' => 'SA2!n2!n18!c'
  },
  'SE' => {
    'country' => 'Sweden',
    'iban_length' => '24',
    'iban_structure' => 'SE[0-9]{2}[0-9]{3}[0-9]{16}[0-9]{1}',
    'is_sepa' => 1,
    'pattern' => 'SE2!n3!n16!n1!n'
  },
  'SI' => {
    'country' => 'Slovenia',
    'iban_length' => '19',
    'iban_structure' => 'SI[0-9]{2}[0-9]{5}[0-9]{8}[0-9]{2}',
    'is_sepa' => 1,
    'pattern' => 'SI2!n5!n8!n2!n'
  },
  'SK' => {
    'country' => 'Slovak Republic',
    'iban_length' => '24',
    'iban_structure' => 'SK[0-9]{2}[0-9]{4}[0-9]{6}[0-9]{10}',
    'is_sepa' => 1,
    'pattern' => 'SK2!n4!n6!n10!n'
  },
  'SM' => {
    'country' => 'San Marino',
    'iban_length' => '27',
    'iban_structure' => 'SM[0-9]{2}[A-Z]{1}[0-9]{5}[0-9]{5}[A-Za-z0-9]{12}',
    'is_sepa' => 0,
    'pattern' => 'SM2!n1!a5!n5!n12!c'
  },
  'TN' => {
    'country' => 'Tunisia',
    'iban_length' => '24',
    'iban_structure' => 'TN59[0-9]{2}[0-9]{3}[0-9]{13}[0-9]{2}',
    'is_sepa' => 0,
    'pattern' => 'TN592!n3!n13!n2!n'
  },
  'TL' => {
    'country' => 'Timor-Leste',
    'iban_length' => '23',
    'iban_structure' => 'TL38[0-9]{3}[0-9]{14}[0-9]{2}',
    'is_sepa' => 0,
    'pattern' => 'TL2!n3!n14!n2!n'
  },
  'TR' => {
    'country' => 'Turkey',
    'iban_length' => '26',
    'iban_structure' => 'TR[0-9]{2}[0-9]{5}[A-Za-z0-9]{1}[A-Za-z0-9]{16}',
    'is_sepa' => 0,
    'pattern' => 'TR2!n5!n1!c16!c'
  },
  'VG' => {
    'country' => 'Virgin Islands, British',
    'iban_length' => '24',
    'iban_structure' => 'VG[0-9]{2}[A-Z]{4}[0-9]{16}',
    'is_sepa' => 0,
    'pattern' => 'VG2!n4!a16!n'
  },
  'XK' => {
    'country' => 'Republic of Kosovo',
    'iban_length' => '20',
    'iban_structure' => 'XK[0-9]{2}[0-9]{4}[0-9]{10}[0-9]{2}',
    'is_sepa' => 0,
    'pattern' => 'XK2!n4!n10!n2!n'
  }
);

sub iban_db {
    lock_hash(%iban_db);
    return \%iban_db;
}

my %lettermap = do {
    my $i = 10;
    map +($_ => $i++), 'A'..'Z';
};

sub numify_iban {
    my ($iban) = @_;

    my $to_check = substr($iban, 4) . substr($iban, 0, 4);
    $to_check =~ s/([A-Za-z])/$lettermap{uc($1)}/g;

    return $to_check;
}

sub mod97 {
    my ($number) = @_;

    # Max 9 digits, safe for 32bit INT
    my ($r, $l) = (0, 9);
    while ($number =~ s/^([0-9]{1,$l})//) {
        $r = $r . $1;
        $r %= 97;
        $l = 9 - length($r);
    }
    return $r;
}

1;

=head1 NAME

Business::IBAN::Database - Simple database for checking IBANs

=head1 SYNOPSIS

    use Business::IBAN::Database;

    my $iso3166a2 = uc substr $iban, 0, 2;
    if (!exists iban_db->{$iso3166a2}) {
        die "Countrycode '$iso3166a2' not in IBAN.\n";
    }
    if (length($iban) != iban_db->{$iso3166a2}{iban_length}) {
        die "Invalid length for '$iban'.\n";
    }
    if ($iban !~ iban_db->{$iso3166a2}{iban_structure}) {
        die "Invalid pattern for '$iban'.\n";
    }
    if (mod97(numify_iban($iban)) != 1) {
        die "Invalid checksum for '$iban'.\n";
    }

=head1 DESCRIPTION

This module was originally generated from the IBAN_Registry.pdf document
supplied by SWIFT version 45 (April 2013) and updated to version 54
(January 2015).

All functions are exported by default.

=head2 iban_db()

Returns a reference to the "database" of known IBAN entities, keyed on the
two letter code for participating countries (See ISO 3166 alpha 2 codes).

=head2 numify_iban($iban)

Put the first four characters at the end of the string. Transform all letters
into numbers. This results in a string of digits [0-9] that can be used as a
number.

=head2 mod97($number)

Returns the remainder of division by 97.

=head1 COPYRIGHT

E<copy> MMXIII-MMXV - Abe Timmerman <abeltje@cpan.org>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut
